Wiki/Developer/ImageMatching.md

```diff
@@ -0,0 +1,185 @@
+
+Waterfill selects for objects on the screen that are a particular color. These Waterfilled objects are then compared with a template image. If the Waterfilled object is similar to the template image, the Detector returns true.
+
+There are several detector/image matchers that you can look at for examples, such as `MapPokeCenterIconDetector` or `FastTravelDetector`. You can also use `GameName_ObjectNameDetector` as a template to build off of.
+
+### Visualizing Waterfill
+
+Suppose we have the following image and we want to detect the Pokecenter. 
+
+<img src="images/image-to-detect.png">
+
+Next, we use Waterfill to get the red portions of the image. The Waterfill algorithm first applies the red color filter, and you're left with something like the following binary image.
+
+<img src="images/binary-image.png">
+
+Waterfill then extracts all connected red portions (that are above a certain area threshold). Note that the red arc and dot inside the icon aren't connected to the rest of the icon, so these won't be a part of the Waterfill object.
+
+<img src="images/waterfill-obj1.png">
+<img src="images/waterfill-obj2.png">
+<img src="images/waterfill-obj3.png">
+
+These Waterfill objects are then compared with the waterfilled template. 
+
+<img src="images/template-image-waterfill.png">
+
+In particular, their area ratio and aspect ratio are compared with the template. Area ratio and aspect ratio are discussed later.
+
+> Note: the binary image and waterfill images above are just for illustrative purposes. In reality, the binary image is actually a binary matrix. Similarly, Waterfill objects are not images, but are actually a class which contains the objects' coordinates on the image, its bounding box, and the number of bits in the object.
+
+If the Waterfill object meets the area ratio and aspect ratio requirements, each Waterfill object will produce a cropped image, where the original image will be cropped according to the bounding box of the Waterfill object. 
+
+<img src="images/waterfill-subimage1.png">
+<img src="images/waterfill-subimage2.png">
+<img src="images/waterfill-subimage3.png">
+
+Each cropped image will then be compared against the cropped template image.
+
+<img src="images/template-image.png">
+
+In particular, RMSD will be computed for each cropped image. Any object that meets the RMSD threshold will be considered a match. 
+
+Note that because we're no longer looking at the Waterfill object, but the cropped image, the red arc and dot inside the icon once again reappear. In other words, the red arc and dot aren't relevant for meeting the aspect ratio and area ratio requirements, but they are relevant for meeting the RMSD threshold.
+
+### Waterfill
+
+Waterfill is an algorithm that filters for all pixels that meet the given color filter(s). It then finds all the remaining pixel groups that are connected and are above a minimum area. Each of these connected pixel groups are considered Waterfill objects.
```

Comment:
Probably worth mentioning that waterfill is computationally cheap and works well with video games with lots of solid colors. General object finding will require machine learning which is both much more computationally intensive and is currently out-of-scope for this project as we don't have the expertise for it yet.